add feature that allows users to request number of comments to display

poncem91 opened the pull request:
Commit lets user select a number of comments to display in the comment history and refreshes the comments loaded with the number selected.

poncem91 committed:
add feature that allows users to request number of comments to display

Commit lets user select a number of comments to display in the comment history and refreshes the comments loaded with the number selected.

poncem91 committed:
Merge branch 'master' into Mafe_AdditionalFeatures

poncem91 committed:
change select number of comments position

Commit moves the select tag to choose the number of comments to display to right above the previous comments.

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,13 +74,14 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     // Builds comment message
-    if (name == null || name.equals("")) {
+    if (name.isEmpty()) {
```
Ooo, this is nicer.

poncem91 replied:
Thanks! :)

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -29,20 +29,35 @@
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
+    String maxCommentsString = request.getParameter("numcomments");
```
We might as well align the naming of the query parameter and what we're calling it here in the servlet. Personally, I like the "max comments" phrasing.

poncem91 replied:
That's a good idea, will change this!

poncem91 replied:
I was just going over this. By the query parameter did you mean the Query("Comment") part? Since this parameter is the entity kind it has to match the same naming as when the comments are added to datastore, so I'd keep the current phrasing. Or were you referring to the request.getParamater("numcomments") so that it matches the maxComments phrasing?

peap replied:
Oops, I mean the parameter in the HTTP query string: `numcomments`.

poncem91 replied:
Done! :)

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -29,20 +29,35 @@
+    try {
+        maxComments = Integer.parseInt(maxCommentsString);
+    } catch (NumberFormatException e) {
+        System.err.println("Could not convert to int: " +  maxCommentsString);
```
What happens to maxComments after we print this error and continue on with the method?

poncem91 replied:
I initially added this when I had initialized maxComments to 10  as the default number of comments but I changed this when I added an argument to the getComments(numComments) function in javascript and instead used the `<body onload="getComments(10)">`line in index.html to set the default number of comments.
When maxComments was initialized to 10, maxComments would keep this value if there was a NumberFormatException.
Now that you ask this I'm wondering whether I need to include this try-catch statement at all if the String maxCommentsString comes directly from the options in the dropdown menu, so the catch might never execute.
If it's better to keep the try-catch statement regardless I could have the catch statement assign 0 to maxComments.

Are there any best practices design-wise for this situation?

peap replied:
Best practice: never trust data from the user :) 

The select box has limited options, but it's easy to manipulate them with, e.g., the inspector, or to intercept the `fetch` requests that the browser makes (on behalf of the user, after all) and modify them. So you always want to cleanse user input the best you can, such as doing try-catch here.

So, I wasn't able to create an exception in my testing, but yes, I would set a default value in the catch block if `parseInt` fails. Probably set it to your default of 10, though, rather than zero, so that the handler fails gracefully and still does something useful.

poncem91 replied:
Sounds good! I'll set it to the default value - I might change this to 5 now that the comment boxes take up more space

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -29,20 +29,35 @@
+    int currentCount = 0;
+
     for (Entity entity : results.asIterable()) {
+      if (currentCount >= maxComments) break;
```
Can we make Datastore do the work of limiting the number of comments returned, instead of tracking it ourselves here? It's better to get only the data we need from the database.

poncem91 replied:
Sounds good! I'll read the documentation to see how to achieve this and change it.

poncem91 committed:
limits datastore results to number of maxComments

Commit makes use of datastore's FetchOptions to set the limit for the query to maxComments. This limits the number of returned comments.

poncem91 committed:
change max number of comments parameter to maxcomments

Commit changes the request parameter for max number of comments from numcomments to maxcomments to align it with the phrasing used in the servlet.

poncem91 committed:
initialize maxComments to 0 if catch is executed

Commit forces maxComments to equal 0 if a NumberFormatException is found when parsing the maxCommentsString input.

peap reviewed (approved):
Approved, with just the one outstanding change about what to do in that try-catch block.

poncem91 committed:
make maxComments default to 5 if there's a NumberFormatException

Commit assigns the value of 5 to maxComments if the try catch statement catches a NumberFormatException when trying to parse maxCommentsString.
